forage-android/src/main/java/com/joinforage/forage/android/core/telemetry/Metrics.kt

```diff
@@ -0,0 +1,204 @@
+    private var vaultType: VaultType? = null
+    private var vaultAction: ActionType? = null
+    private var responseType: ResponseType = ResponseType.VAULT_RESPONSE_TIME
+
+    init {
+        this.vaultType = vault
+        this.vaultAction = vaultAction
+    }
+
+    internal companion object {
+        internal fun newMeasurement(vault: VaultType, vaultAction: ActionType, metricsLogger: Log?): VaultProxyResponseMonitor {
+            return VaultProxyResponseMonitor(vault, vaultAction, metricsLogger)
+        }
+    }
+
+    override fun logWithResponseAttributes(
+        metricsLogger: Log?,
+        responseAttributes: Map<String, Any>
+    ) {
+        val path = responseAttributes[MetricsConstants.PATH]
+        val method = responseAttributes[MetricsConstants.METHOD]
+        val httpStatus = responseAttributes[MetricsConstants.HTTP_STATUS]
+        val responseTime = responseAttributes[MetricsConstants.RESPONSE_TIME_MS]
+
+        if (path == null || method == null || httpStatus == null || responseTime == null) {
+            metricsLogger?.e("[Metrics] Incomplete or missing response attributes. Could not log metric.")
+            return
+        }
+
+        val vaultType = vaultType
+        val action = vaultAction
+
+        metricsLogger?.i(
+            "[Metrics] Received response from $vaultType proxy",
+            attributes = mapOf(
+                MetricsConstants.PATH to path,
+                MetricsConstants.METHOD to method,
+                MetricsConstants.HTTP_STATUS to httpStatus,
+                MetricsConstants.RESPONSE_TIME_MS to responseTime,
+                MetricsConstants.VAULT_TYPE to vaultType,
+                MetricsConstants.ACTION to action,
+                MetricsConstants.RESPONSE_TYPE to responseType
+            )
+        )
+    }
+}
+
+internal class CustomerPerceivedResponseMonitor(vault: VaultType, vaultAction: ActionType, metricsLogger: Log?) : ResponseMonitor(metricsLogger) {
+    private var vaultType: VaultType? = null
+    private var vaultAction: ActionType? = null
+    private var responseType: ResponseType = ResponseType.CUSTOMER_PERCEIVED_RESPONSE_TIME
+
+    init {
+        this.vaultType = vault
+        this.vaultAction = vaultAction
+    }
+
+    internal companion object {
+        internal fun newMeasurement(vault: VaultType, vaultAction: ActionType, metricsLogger: Log?): CustomerPerceivedResponseMonitor {
+            return CustomerPerceivedResponseMonitor(vault, vaultAction, metricsLogger)
+        }
+    }
+
+    override fun logWithResponseAttributes(
+        metricsLogger: Log?,
+        responseAttributes: Map<String, Any>
+    ) {
+        val responseTime = responseAttributes[MetricsConstants.RESPONSE_TIME_MS]
+
+        if (responseTime == null) {
+            metricsLogger?.e("[Metrics] Incomplete or missing response attributes. Could not log metric.")
+            return
+        }
```

Comment:
we should assert that `EventName` (or RESPONSE_TYPE) and `action` also set to the right EventName here

to ensure our metrics are tuned correctly and to help us distinguish between BIs + captures

```kotlin
if (responseTime == null || eventName != CUSTOMER_PERCEIVED_RESPONSE || action == null) {
...
}
```

Reply:
`vaultAction` and `vaultType` can't be null and we can't validate their value at this point.

Reply:
sounds good, i've amended my comment^